akka-stream-typed/src/main/scala/akka/stream/typed/scaladsl/ActorFlow.scala

```diff
@@ -0,0 +1,136 @@
+  def ask[I, Q, O](ref: ActorRef[Q])(makeMessage: (I, ActorRef[O]) ⇒ Q)(implicit timeout: Timeout): Flow[I, O, NotUsed] =
+    ask(parallelism = 2)(ref)(makeMessage)(timeout)
+
+  /**
+   * Use the `ask` pattern to send a request-reply message to the target `ref` actor.
+   * If any of the asks times out it will fail the stream with a [[java.util.concurrent.TimeoutException]].
+   *
+   * Do not forget to include the expected response type in the method call, like so:
+   *
+   * {{{
+   * flow.ask[String, Asking, Reply](parallelism = 4)(ref, (el, replyTo) => Asking(el, replyTo))
+   *
+   * // or even:
+   * flow.ask[String, Asking, Reply](parallelism = 4)(ref, Asking(_, _))
+   * }}}
+   *
+   * otherwise `Nothing` will be assumed, which is most likely not what you want.
+   *
+   * The mapTo class parameter is used to cast the incoming responses to the expected response type.
+   *
+   * Similar to the plain ask pattern, the target actor is allowed to reply with `akka.util.Status`.
+   * An `akka.util.Status#Failure` will cause the stage to fail with the cause carried in the `Failure` message.
+   *
+   * The stage fails with an [[akka.stream.WatchedActorTerminatedException]] if the target actor is terminated,
+   * or with an [[java.util.concurrent.TimeoutException]] in case the ask exceeds the timeout passed in.
+   *
+   * Adheres to the [[ActorAttributes.SupervisionStrategy]] attribute.
+   *
+   * '''Emits when''' the futures (in submission order) created by the ask pattern internally are completed
+   *
+   * '''Backpressures when''' the number of futures reaches the configured parallelism and the downstream backpressures
+   *
+   * '''Completes when''' upstream completes and all futures have been completed and all elements have been emitted
+   *
+   * '''Fails when''' the passed in actor terminates, or a timeout is exceeded in any of the asks performed
+   *
+   * '''Cancels when''' downstream cancels
+   *
+   * @tparam I Incoming element type of the Flow
+   * @tparam Q Question message type that is spoken by the target actor
+   * @tparam O answer type that the Actor is expected to reply with, it will become the Output type of this Flow
+   */
+  @implicitNotFound("Missing an implicit akka.util.Timeout for the ask() stage")
+  def ask[I, Q, O](parallelism: Int)(ref: ActorRef[Q])(makeMessage: (I, ActorRef[O]) ⇒ Q)(implicit timeout: Timeout): Flow[I, O, NotUsed] = {
+    import akka.actor.typed.scaladsl.adapter._
+    val untypedRef = ref.toUntyped
+
+    val askFlow = Flow[I]
+      .watch(untypedRef)
+      .mapAsync(parallelism) { el ⇒
+        akka.pattern.extended.ask(untypedRef, (replyTo: akka.actor.ActorRef) ⇒ makeMessage(el, replyTo))
+      }
+      .map(_.asInstanceOf[O])
+      .recover[O] {
+        case ex: AskTimeoutException ⇒
+          // in Akka Typed we use the `TimeoutException` everywhere
+          throw new java.util.concurrent.TimeoutException(ex.getMessage)
```

Comment:
according to docs this will be logged at ERROR level, which we might not want. use `mapError` instead?

Reply:
I wanted them to be loud intentionally bot OK to move to mapError I suppose, done

Reply:
it will fail the stream, which is *the* way to signal errors, we shouldn't add logging in random places